src/daft-session/src/session.rs

```diff
@@ -0,0 +1,119 @@
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard},
+};
+
+use daft_catalog::{catalog::Catalogs, DataCatalog, Identifier, Name};
+use daft_logical_plan::LogicalPlanBuilder;
+
+// use options::Options;
+use crate::{
+    error::Result, obj_already_exists_err, obj_not_found_err, options::Options, unsupported_err,
+};
+
+/// Session holds all state for query planning and execution (e.g. connection).
+#[derive(Debug, Clone)]
+pub struct Session {
+    /// Session state for interior mutability
+    state: Arc<RwLock<SessionState>>,
+}
+
+/// Session state is to be kept internal, consider a builder.
+#[derive(Debug)]
+struct SessionState {
+    /// Session identifier
+    _id: String,
+    /// Session options i.e. curr_catalog and curr_schema.
+    _options: Options,
+    /// References to the available catalogs.
+    catalogs: Catalogs,
+    // TODO execution context
+    // TODO temporary! tables come from a catalog, not here!!
+    tables: HashMap<String, LogicalPlanBuilder>,
+}
+
+impl Session {
+    /// Creates a new empty session
+    pub fn new(id: &str) -> Self {
+        let state = SessionState {
+            _id: id.to_string(),
+            _options: Options::default(),
+            catalogs: Catalogs::empty(),
+            tables: HashMap::new(),
+        };
+        let state = RwLock::new(state);
+        let state = Arc::new(state);
+        Self { state }
+    }
```

Comment:
nit: I'd make the `id` optional as sometimes we'll just want a session without caring what the id should be. 

```rs
pub fn new() -> Self {
  let id = Uuid::new_v4().to_string();
  Self::new_with_id(id)
}
pub fn new_with_id(id: &str) -> Self {
}
```

Reply:
Ok, I have a TODO for a builder so I updated the session id like you've suggested and won't introduce any additional constructors until we have a use-case for the builder – which will be added to python as well.